### PR TITLE
catalog: emit all common relations based on specs

### PR DESCRIPTION
**Pull request by freben**
Start emitting all known relation types from the core entity kinds, based on their spec data.

This all happens as a post processing step in the `BuiltinKindsEntityProcessor`.

### CODE REVIEW

File: packages/catalog-model/src/entity/ref.ts [hunk 1]
```diff
@@ -86,25 +86,25 @@ export function parseEntityName(
  */
 export function parseEntityRef(
   ref: EntityRef,
-  context?: { defaultKind: string },
```
Comment (freben) [inline packages/catalog-model/src/entity/ref.ts:89]: these are just a reordering of the signatures, so the most precise is at the top; otherwise TS sometimes picks the wrong one

File: packages/catalog-model/src/kinds/relations.ts [hunk 2]
```diff
@@ -33,7 +33,9 @@ export const RELATION_OWNER_OF = 'ownerOf';
  * A relation with an API entity, typically from a component or system
  */
 export const RELATION_CONSUMES_API = 'consumesApi';
+export const RELATION_API_CONSUMED_BY = 'apiConsumedBy';
 export const RELATION_PROVIDES_API = 'providesApi';
+export const RELATION_API_PROVIDED_BY = 'apiProvidedBy';
```
Comment (freben) [inline packages/catalog-model/src/kinds/relations.ts:38]: this seems like an omission on our part - we mixed up the two concepts of providing and consuming

**Comment by freben** [comment]
@Fox32 this may be interesting to you

File: plugins/catalog-backend/migrations/20201123205611_relations_table_uniq.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        .notNullable()
+        .comment('The full name of the target entity of the relation');
+      table.index('originating_entity_id', 'originating_entity_id_idx');
+      table.index('source_full_name', 'source_full_name_idx');
+    });
+  } else {
+    await knex.schema.alterTable('entities_relations', table => {
+      table.dropPrimary();
+      table.index('originating_entity_id', 'originating_entity_id_idx');
+      table.index('source_full_name', 'source_full_name_idx');
```
Comment (Rugvip) [inline plugins/catalog-backend/migrations/20201123205611_relations_table_uniq.js:54]: What's the purpose of this change? Why have and index for the originating entity ID, and why remove the indexing of the type and target full name?

**Comment by Fox32** [comment]
> @Fox32 this may be interesting to you

@freben  Can you read my mind 😱 ? I just thought about it some minutes ago an looked up what relations were already exported and noticed that some are missing 😆 

Very cool! Looking forward to integrate API providers and consumers once #2987 is completed!

File: packages/catalog-model/src/kinds/relations.ts (hunk 2 above)
Comment (Rugvip) [reply packages/catalog-model/src/kinds/relations.ts:38]: No provides/consumes are the two different directions 😅, what's the difference between `apiProvidedBy` and `consumesApi`? :grin:

File: plugins/catalog-backend/migrations/20201123205611_relations_table_uniq.js (hunk 3 above)
Comment (freben) [reply plugins/catalog-backend/migrations/20201123205611_relations_table_uniq.js:54]: Ah the originating entity ID is a mistake - it should already be indexed by virtue of being an FK constraint. Fixing.

As for the other one, the source name is the [primary means of joining / lookup](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/CommonDatabase.ts#L504). The others aren't used in a fashion that requires an index, and I just didn't want to have too many indices for insert-performance reasons. For lookups, the current system at least is to use the `entities_search` table to find matches, not other indexed lookups.

Removing the compound constraint is due to local processing constraints - a group A that says it has a child B, while B says it has parent A, will emit duplicate parent/child relations.

File: packages/catalog-model/src/kinds/relations.ts (hunk 2 above)
Comment (freben) [reply packages/catalog-model/src/kinds/relations.ts:38]: C1 <-> A <-> C2

- `A.apiProvidedBy.C1` + `C1.providesApi.A`: `C1` has a listening socket that you can make `A` requests on
- `C2.consumesApi.A` + `A.apiConsumedBy.C2`: `C2` makes fetch requests to services that provide `A` (possibly `C1`)

Comment (Rugvip) [reply packages/catalog-model/src/kinds/relations.ts:38]: Ah yeah did not consider that there's an API inbetween, you're totally right x)

Comment (Fox32) [reply packages/catalog-model/src/kinds/relations.ts:38]: I think we might have to rename the `implementsApis` in the component entity in the future. A better name would be `providesApis`? This would be a breaking change similar to the ancestors and descendants?

Later we can add the same for `consumesApis`. We can also support the "complex" syntax for APIs that work without manually defining an entity, but that is something we can introduce later without a breaking change (see https://github.com/backstage/backstage/pull/2840#issuecomment-713835013)

Comment (freben) [reply packages/catalog-model/src/kinds/relations.ts:38]: Mm, that's a breaking change for sure, but we could support both in the time between deprecation and deletion. You suggest this mainly from a language consistency standpoint right?

**Comment by freben** [comment]
OK, removed the extra index line in `down` too @Rugvip

File: packages/catalog-model/src/kinds/relations.ts (hunk 2 above)
Comment (Fox32) [reply packages/catalog-model/src/kinds/relations.ts:38]: Supporting both should be easy. Yes, for consistency only

**Review by benjdlambert — APPROVED** [review]
